third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/Allocation.cpp

```diff
@@ -0,0 +1,83 @@
+#include <memory>
+
+#include "Allocation.h"
+#include "TargetInfo.h"
+#include "TritonNVIDIAGPUToLLVM/Passes.h"
+#include "triton/Analysis/Allocation.h"
+#include "triton/Conversion/TritonGPUToLLVM/AllocateSharedMemoryUtility.h"
+#include "triton/Dialect/Triton/IR/Utility.h"
+#include "triton/Tools/GenericSwizzling.h"
+#include "triton/Tools/LayoutUtils.h"
+
+using namespace mlir;
+using namespace mlir::triton;
+
+namespace mlir {
+namespace triton {
+#define GEN_PASS_DEF_ALLOCATESHAREDMEMORYNV
+#include "TritonNVIDIAGPUToLLVM/Passes.h.inc"
+} // namespace triton
+} // namespace mlir
+
+namespace {
+static unsigned getNumScratchElemsSwizzledCvt(RankedTensorType srcTy,
+                                              RankedTensorType dstTy,
+                                              TargetInfoBase &targetInfo) {
+  auto *ctx = srcTy.getContext();
+  auto srcLayout = triton::gpu::toLinearLayout(srcTy);
+  auto dstLayout = triton::gpu::toLinearLayout(dstTy);
+  srcLayout = actionRemoveBroadcastedRegs(srcLayout).apply(srcLayout);
+  dstLayout = actionRemoveBroadcastedRegs(dstLayout).apply(dstLayout);
+  auto bitwidth = getBitwidth(srcTy);
+  auto [smem, _] =
+      triton::gpu::optimalSwizzling(srcLayout, dstLayout, bitwidth, targetInfo);
+  auto reps = smem.getInDimSize(StringAttr::get(ctx, "reps"));
+  return smem.getTotalOutDimSize() / reps;
+}
+
+struct AllocateSharedMemoryNv
+    : public mlir::triton::impl::AllocateSharedMemoryNvBase<
+          AllocateSharedMemoryNv> {
+  using AllocateSharedMemoryNvBase::AllocateSharedMemoryNvBase;
+
+  AllocateSharedMemoryNv(int32_t computeCapability, int32_t ptxVersion)
+      : AllocateSharedMemoryNvBase({computeCapability, ptxVersion}) {}
+
+  void runOnOperation() override {
+    ModuleOp mod = getOperation();
+    mlir::triton::NVIDIA::TargetInfo targetInfo(computeCapability, ptxVersion);
+    ModuleAllocation allocation(
+        mod, mlir::triton::nvidia_gpu::getNvidiaAllocationAnalysisScratchSizeFn(
+                 targetInfo));
+    mlir::triton::gpu::attachAllocationSizeAndOffsetAttr(mod, allocation);
+  }
+};
+} // namespace
+
+namespace mlir::triton::nvidia_gpu {
+std::function<unsigned(Operation *)>
+getNvidiaAllocationAnalysisScratchSizeFn(TargetInfoBase &targetInfo) {
```

Comment:
Can you just override the default allocation? AMD has their own allocation pass as well.

Reply:
I think the code here is more general and correct than what we have in `Allocation.cpp`.

Reply:
The issue is this returns the shape needed if you can use ldmatrix/stmatrix, which is not generally the case... In fact, it is dependent on the instructions that you can use (on your `targetInfo`). That's why it has its own pass.